### PR TITLE
Autoencoder Conv3D for bias correction

### LINKED ISSUE #2: Design a ML model for bias correction
In this task, three ML platforms (Pytorch, Tensorflow and Keras) will be investigated for GFSv16 post product bias correction. One will be chosen for this project with best extensibility, user friendly environment and potential for operational use. Also, a Machine Learning method will be designed and developed for bias correction.

### PR DESCRIPTION
**Pull request by sadeghitabas**
## Description
In this PR, three ML platforms (Pytorch, Tensorflow and Keras) investigated for GFSv16 post-product bias correction and Pytorch was selected due to its extensibility, user friendly environment and potential for operational use. Also, a Machine Learning method, an Autoencoder Conv3D model, called ML4BC developed and designed for GFS 2m temperature bias correction. The model training inputs were prepared using `preprocessing.py` from GFS (biased) and ERA5 (unbiased) data from 20210321 to 20231018 in every 6 hours. The data has 0.25-degree spatial resolution and 50 hourly timesteps (e.g., [721,1440,50]). 

In summary, this model provides functionalities, including:
- `preprocessing.py`: preprocessing.py: a utility for preparing ML4BC model inputs from 0.25-degree resolution GFS and ERA5 data. This Python utility has two modules one for GFS info and the second one for era5 info. 
- `autoencoder_model.py`: Main model structure.
- `netcdf_dataset.py`: Provide functionalities for data processing, normalizing, rescaling and making pytorch dataloader for both GFS and ERA5.
- `check_missing_files`: A function for checking missing files.
- `calculate_mean_and_std`: A function to calculate mean and standard deviation of training dataset which provides values for normalization and rescaling modules. 
- `ml4bc.py`: Model Initiation, Training Loop, Module for Saving Model State
- `ml4bc.ipynb`: An example of the ML4BC modeling process.
- `postprocessing.ipynb`: A notebook for postprocessing including data sanity check, plotting, and data analysis.

### Associated GraphCast Issue to close
- Closes https://github.com/NOAA-EMC/ML4BC/issues/2

### Input data
- [x] No changes are expected to input data.
- [ ] Changes are expected to input data:
  - [ ] New input data.
  - [ ] Updated input data.

### Needed libraries

  - `python3`
  - `torch`
  - `torchvision`
  - `torchaudio`
  - `CUDA v12`
  - `netCDF4`
  - `numpy`
  - `matplotlib`
  - `tqdm`

### CODE REVIEW

**Comment by sadeghitabas** [comment]
Dear @junwang-noaa, this PR is ready to review. Thanks

File: autoencoder_model.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+'''
+Description: This script provides the main autoencoder model structure that is used for GFS 2m temperature bias correction task.
```
Comment (junwang-noaa) [inline autoencoder_model.py]: I am thinking this can be applied to other products too (e.g. precip). Can we change GFS 2m temperature to GFS forecast products?

File: ml4bc.ipynb [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    "\n",
+    "    def __getitem__(self, idx):\n",
+    "        file_path = os.path.join(self.root_dir, self.file_list[idx])\n",
+    "        \n",
+    "        # Load NetCDF data\n",
+    "        dataset = nc.Dataset(file_path)\n",
+    "        data = dataset.variables['t2m'][:].astype(np.float32)  # Adjust 'data' to the variable name in your file\n",
+    "        dataset.close()\n",
+    "        \n",
+    "        # Reshape the data to (1, 50, 721, 1440)\n",
```
Comment (junwang-noaa) [inline ml4bc.ipynb:141]: Can we make "50" as an argument that can be set up at run time?

File: postprocessing.ipynb [hunk 3]
```diff
@@ -0,0 +1,1296 @@
...
+       "       &#x27;2021-03-23T02:00:00.000000000&#x27;, &#x27;2021-03-23T03:00:00.000000000&#x27;,\n",
```
Comment (junwang-noaa) [inline postprocessing.ipynb:634]: May I ask if this code is generated automatically? I see the date here is very specific. How do we update the date if we are using a different period of time?

File: ml4bc.ipynb (hunk 2 above)
Comment (sadeghitabas) [reply ml4bc.ipynb:141]: Its not like other models, first we need to make the dataset in that format. Then we can increase the channel. Once a model is trained you cannot do any changes to the structure, otherwise you need to train it again. I expect my training process would be done by this weekend. Tomorrow, we can talk about our future directions and updates.

File: autoencoder_model.py (hunk 1 above)
Comment (sadeghitabas) [reply autoencoder_model.py]: Updated!
Yes, it can be applied; however, we need to train it again.

File: postprocessing.ipynb (hunk 3 above)
Comment (sadeghitabas) [reply postprocessing.ipynb:634]: This is a Jupyter notebook file, and you need to open it using Jupyter. It can easily be updated to an arbitrarily chosen date. We can talk about it tomorrow.

File: ml4bc.ipynb (hunk 2 above)
Comment (junwang-noaa) [reply ml4bc.ipynb:141]: Ok, thanks

**Review by junwang-noaa — APPROVED** [review]
